File: src/main/java/no/nav/syfo/exception/RestTemplateErrorHandler.java
package no.nav.syfo.exception;

import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.http.client.ClientHttpResponse;
import org.springframework.web.client.ResponseErrorHandler;

import java.io.IOException;
import java.net.URI;

@Slf4j
public class RestTemplateErrorHandler implements ResponseErrorHandler {
    @Override
    public boolean hasError(ClientHttpResponse httpResponse) throws IOException {
        switch (httpResponse.getStatusCode().series()) {
            case SUCCESSFUL:
                return false;
            case CLIENT_ERROR:
                return httpResponse.getStatusCode() != HttpStatus.FORBIDDEN;
            default:
                return true;
        }
    }

    @Override
    public void handleError(ClientHttpResponse httpResponse) throws IOException {}

    @Override
    public void handleError(URI url, HttpMethod method, ClientHttpResponse httpResponse) throws IOException {
        String requestUrlWithHiddenFnr = hideFnrFromUrl(url.toString());
        if (httpResponse.getStatusCode().series() == HttpStatus.Series.SERVER_ERROR) {
            log.error("Fikk server error ved {}-kall til {}. statusCode: {}, statusText: {}, body: {}", method, requestUrlWithHiddenFnr, httpResponse.getStatusCode(), httpResponse.getStatusText(), httpResponse.getBody());
            throw new RuntimeException("Fikk en server-error ved " + method + "-kall til " + requestUrlWithHiddenFnr);
        } else if (httpResponse.getStatusCode().series() == HttpStatus.Series.CLIENT_ERROR) {
            log.error("Fikk en client error ved {}-kall til {}, som ikke er forbidden. statusCode: {}, statusText: {}, body: {}", method, requestUrlWithHiddenFnr, httpResponse.getStatusCode(), httpResponse.getStatusText(), httpResponse.getBody());
            throw new RuntimeException("Noe gikk galt ved " + method + "-kall til " + requestUrlWithHiddenFnr);
        }
    }

    private String hideFnrFromUrl(String url) {
        return url.replaceAll("\\d", "*");
    }
}

File: src/main/kotlin/no/nav/syfo/api/exception/RestTemplateErrorHandler.kt
package no.nav.syfo.api.exception

import org.slf4j.LoggerFactory
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatus.Series
import org.springframework.http.client.ClientHttpResponse
import org.springframework.web.client.ResponseErrorHandler
import java.io.IOException
import java.net.URI

class RestTemplateErrorHandler : ResponseErrorHandler {
    @Throws(IOException::class)
    override fun hasError(httpResponse: ClientHttpResponse): Boolean {
        return when (httpResponse.statusCode.series()) {
            Series.SUCCESSFUL -> false
            Series.CLIENT_ERROR -> httpResponse.statusCode != HttpStatus.FORBIDDEN
            else -> true
        }
    }

    @Throws(IOException::class)
    override fun handleError(httpResponse: ClientHttpResponse) {
    }

    @Throws(IOException::class)
    override fun handleError(url: URI, method: HttpMethod, httpResponse: ClientHttpResponse) {
        val requestUrlWithHiddenFnr = hideFnrFromUrl(url.toString())
        if (httpResponse.statusCode.series() == Series.SERVER_ERROR) {
            log.error("Fikk server error ved {}-kall til {}. statusCode: {}, statusText: {}, body: {}", method, requestUrlWithHiddenFnr, httpResponse.statusCode, httpResponse.statusText, httpResponse.body)
            throw RuntimeException("Fikk en server-error ved $method-kall til $requestUrlWithHiddenFnr")
        } else if (httpResponse.statusCode.series() == Series.CLIENT_ERROR) {
            log.error("Fikk en client error ved {}-kall til {}, som ikke er forbidden. statusCode: {}, statusText: {}, body: {}", method, requestUrlWithHiddenFnr, httpResponse.statusCode, httpResponse.statusText, httpResponse.body)
            throw RuntimeException("Noe gikk galt ved $method-kall til $requestUrlWithHiddenFnr")
        }
    }

    private fun hideFnrFromUrl(url: String): String {
        return url.replace("\\d".toRegex(), "*")
    }

    companion object {
        private val log = LoggerFactory.getLogger(RestTemplateErrorHandler::class.java)
    }
}
